### PR TITLE
feat: [FC-0047] xBlock offline mode

### PR DESCRIPTION
**Pull request by IvanStepanok**
# 🛜 Offline mode – Quick Preview
### Downloading

https://github.com/openedx/openedx-app-ios/assets/128456094/5e5b0b89-3c1e-4c3f-a1fa-b5cbea7a1036

### Working with blocks without an internet connection

https://github.com/openedx/openedx-app-ios/assets/128456094/58d1e162-10cb-477a-ab0a-b9d0616ff4e5

### Offline content is synchronized immediately after connecting to the network

https://github.com/openedx/openedx-app-ios/assets/128456094/b79bcc58-d551-474a-8a53-39ed001913e3

### Environment for testing:
```
API_HOST_URL: 'https://axim-mobile.raccoongang.net'
ENVIRONMENT_DISPLAY_NAME: 'lms-axim-stage'
FEEDBACK_EMAIL_ADDRESS: 'support@example.com'
OAUTH_CLIENT_ID: 'zP3vPz00c8fTRpYjNbVSlA1fxt9LnCxTM4JK1KQ0'
DISCOVERY:
    TYPE: "native"
    
FIREBASE:
    ENABLED: false

UI_COMPONENTS:
    COURSE_UNIT_PROGRESS_ENABLED: false
    COURSE_NESTED_LIST_ENABLED: false

```

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @IvanStepanok! Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

**Comment by rnr** [comment]
@IvanStepanok Thank you for your work on this feature - looks cool!
I have a few comments on the UI - will place here

**Comment by rnr** [comment]
While the app is downloading for offline, the user stops the app and starts it again; after navigating to the course -> offline user sees the ‘Download All’ button, but the download continues.

https://github.com/user-attachments/assets/980da3e1-49b4-4db5-b44e-3c2be83a165c

**Comment by rnr** [comment]
All content in course is loaded. User deletes only one sequence. After tap on "Download all" button it shows like user will get whole course again. 

https://github.com/user-attachments/assets/996b52cd-feff-4b3c-a832-9b3d0234bd52

**Comment by rnr** [comment]
3) All content in course is loaded. Some chapter contains youtube videos only (which are not downloadable). When goes to Videos Tab and try to remove such chapter it shows "Removing this content will free up 0MB" - do we need to show 'downloaded' state for chapters in Videos which don't contain downloaded videos?

**Comment by rnr** [comment]
4) A user has downloaded a course. The user answered a few questions offline, went online, the answers were synchronised, and now we see them in the online version. (it's working right)
When the user goes to that test again offline, he don't see the answers - it looks like the user is seeing the downloaded version BEFORE he gave the answers.

https://github.com/user-attachments/assets/9cb337ac-5146-4f3a-a52d-ce5b43c9c2b7

**Comment by rnr** [comment]
5) The user can't select a previously selected option in an offline quiz. For example: select option 3 - submit - change your mind and select option 4 - submit - change your mind again and try to select option 3 - unable to select option 3 again

https://github.com/user-attachments/assets/161278c2-8c6c-4a34-98b6-cb0a8a3dbd90

**Comment by IvanStepanok** [comment]
Hi @rnr, and thank you for the review! It is very valuable to me to get a view from another angle. I appreciate it.

1. ✅ Fixed

https://github.com/user-attachments/assets/a2deac8f-b047-4e6f-bad4-97f0abf4a105

2. ✅Fixed (Pay attention. We don't show download notification, less than 100 MB when connected to Wi-Fi)

https://github.com/user-attachments/assets/e6fd90a6-535d-4fad-a576-c3e978ca7e34

3. ✅ Fixed

https://github.com/user-attachments/assets/12b5fb70-daad-4859-a101-d0fcf32fae48

I also noticed that if we delete one file, when downloading, the dialog box shows us the size of the entire group instead of a single undownloaded file. This is also fixed.



6. 🤷‍♂️ This is a known issue, the Open EDX team plans to improve this flow in the future. Currently, the backend has IOSBridge only for offline mode. As soon as they finalize the backend, it will work for us immediately.

7. 🤷‍♂️ This is also the behavior that was set on the backend side, on our side we cannot influence it in any way.

**Comment by rnr** [comment]
@IvanStepanok Thank you for the fixes and clarifications
Looks like we have issue for video:
2. It seems that the video file is not counted in the total size, even though the dialogue says that this space will be freed up.

https://github.com/user-attachments/assets/09aac20f-8e1a-4f9c-ad26-630a48391aaf

**Comment by IvanStepanok** [comment]
Hi, @rnr, and thanks for the question. In this case, we don't know the size of a video until we download it because the server returns 0 megabytes. We can measure the actual size of the videos only after downloading. To prevent this behavior, we need to receive the actual video size in the server response.

File: OpenEdX/Data/CorePersistence.swift [hunk 1]
```diff
@@ -242,6 +294,8 @@ public class CorePersistence: CorePersistenceProtocol {
             }
         }
     }
+//<<<<<<< HEAD
```
Comment (rnr) [inline OpenEdX/Data/CorePersistence.swift:297]: unneeded comments?

File: Course/Course/Presentation/Container/CourseContainerViewModel.swift [hunk 2]
```diff
@@ -518,7 +690,94 @@ public class CourseContainerViewModel: BaseCourseViewModel {
...
+    
+    @MainActor
+    func downloadAll() async {
+        guard let course = courseStructure else { return }
+        var blocksToDownload: [CourseBlock] = []
+        var sequentialsToDownload: [CourseSequential] = []
+        
+        for chapter in course.childs {
+            for sequential in chapter.childs where sequential.isDownloadable {
+                var blocks = downloadableBlocks(from: sequential)
```
Comment (rnr) [inline Course/Course/Presentation/Container/CourseContainerViewModel.swift:702]: should be `let`

File: Course/Course/Presentation/Container/CourseContainerViewModel.swift [hunk 3]
```diff
@@ -527,7 +786,75 @@ public class CourseContainerViewModel: BaseCourseViewModel {
...
+            if let fileSize = fileAttributes[.size] as? Int, fileSize > 0 {
+                return fileSize
+            }
+        } catch {
+            debugLog("Error getting file size: \(error.localizedDescription)")
+        }
+        return nil
+    }
+    
+
```
Comment (rnr) [inline Course/Course/Presentation/Container/CourseContainerViewModel.swift:857]: vertical space violation

File: Course/Course/Presentation/Container/CourseContainerViewModel.swift [hunk 4]
```diff
@@ -594,7 +940,139 @@ public class CourseContainerViewModel: BaseCourseViewModel {
...
+        } catch {
+            print("Error retrieving free disk space: \(error.localizedDescription)")
+        }
+        return nil
+    }
+    
+    private func getUsedDiskSpace() -> Int? {
+        do {
+            let attributes = try FileManager.default.attributesOfFileSystem(forPath: NSHomeDirectory() as String)
+            if let totalSpace = attributes[.systemSize] as? Int64, let freeSpace = attributes[.systemFreeSize] as? Int64 {
```
Comment (rnr) [inline Course/Course/Presentation/Container/CourseContainerViewModel.swift:966]: line length violation

File: Course/Course/Presentation/Offline/OfflineView.swift [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+                            .stroke(Theme.Colors.snackbarErrorColor, lineWidth: 2)
+                    )
+                    .background(Theme.Colors.background)
+                    .cornerRadius(8)
+                }
+            }
+            .padding(.vertical, 4)
+        }
+    }
+}
```
Comment (rnr) [inline Course/Course/Presentation/Offline/OfflineView.swift:235]: should we add preview for this view?

File: Course/Course/Presentation/Subviews/ActionViews/DeviceStorageFullAlertView.swift [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+        }
+        .ignoresSafeArea()
+        .onAppear {
+            withAnimation(Animation.linear(duration: 0.3).delay(0.2)) {
+                fadeEffect = true
+            }
+        }
+    }
+    
+
```
Comment (rnr) [inline Course/Course/Presentation/Subviews/ActionViews/DeviceStorageFullAlertView.swift:51]: vertical space violation

File: Course/Course/Presentation/Subviews/ActionViews/DeviceStorageFullAlertView.swift [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+            let contentSize = Double(contentSize)
+            let total = usedSpace + contentSize
+            
+            let minSize: Double = 0.1
+            let usedSpacePercentage = (usedSpace / total) + minSize
+            let contentSizePercentage = (contentSize / total) + minSize
+            let normalizationFactor = 1 / (usedSpacePercentage + contentSizePercentage)
+            
+            let normalizedUsedSpaceWidth = usedSpacePercentage * normalizationFactor
+            let normalizedContentSizeWidth = contentSizePercentage * normalizationFactor
```
Comment (rnr) [inline Course/Course/Presentation/Subviews/ActionViews/DeviceStorageFullAlertView.swift:181]: unused variable

File: Course/Course/Presentation/Subviews/ActionViews/DownloadErrorAlertView.swift [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+                    CourseSequential(
+                        blockId: "",
+                        id: "",
+                        displayName: "Course intro",
+                        type: .chapter,
+                        completion: 0,
+                        childs: [],
+                        sequentialProgress: nil,
+                        due: nil
+                    ),
```
Comment (rnr) [inline Course/Course/Presentation/Subviews/ActionViews/DownloadErrorAlertView.swift:274]: trailing comma violation

File: Course/Course/Presentation/Subviews/CustomDisclosureGroup.swift [hunk 9]
```diff
@@ -213,9 +205,15 @@ struct CustomDisclosureGroup: View {
...
+    private func canDownloadAllSections(in chapter: CourseChapter, videoOnly: Bool) -> Bool {
         for sequential in chapter.childs {
-            if let state = viewModel.sequentialsDownloadState[sequential.id] {
+            if videoOnly {
+                let isDownloadable = sequential.childs.flatMap {
+                    $0.childs.filter({ $0.type == .video })
+                }.contains(where: { $0.isDownloadable })
+                guard isDownloadable else { return false }
+            }
+            if let _ = viewModel.sequentialsDownloadState[sequential.id] {
```
Comment (rnr) [inline Course/Course/Presentation/Subviews/CustomDisclosureGroup.swift:216]: change to `!=nil` please

File: Course/Course/Presentation/Unit/CourseUnitView.swift [hunk 10]
```diff
@@ -413,7 +429,7 @@ public struct CourseUnitView: View {
                 Spacer()
             }
             VStack {
-                if !isHorizontal {
+                if (!isHorizontal) {
```
Comment (rnr) [inline Course/Course/Presentation/Unit/CourseUnitView.swift:432]: Do we need brackets here?

File: Course/Course/Presentation/Unit/Subviews/OfflineContentView.swift [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+            Text(contentState.description)
+                .font(Theme.Fonts.bodyLarge)
+                .multilineTextAlignment(.center)
+                .frame(maxWidth: .infinity)
+                .padding(.top, 12)
+            Spacer()
+        }
+        .padding(24)
+    }
+}
```
Comment (rnr) [inline Course/Course/Presentation/Unit/Subviews/OfflineContentView.swift:61]: Do we need preview for this view?

File: OpenEdX/AppDelegate.swift [hunk 12]
```diff
@@ -195,4 +199,43 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
         guard let deepLinkManager = Container.shared.resolve(DeepLinkManager.self) else { return }
         deepLinkManager.configureDeepLinkService(launchOptions: launchOptions)
     }
+    
+    // Background progress update
+    
+    func registerBackgroundTask() {
+        let isRegistered = BGTaskScheduler.shared.register(forTaskWithIdentifier: Self.bgAppTaskId, using: nil) { task in
```
Comment (rnr) [inline OpenEdX/AppDelegate.swift:206]: line length violation

**Review by rnr — COMMENTED** [review]
Amazing work done. Please fix some small comments

File: Core/Core/View/Base/FileWebView.swift [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+        webview.scrollView.layer.maskedCorners = [.layerMinXMinYCorner, .layerMaxXMinYCorner]
+        
+        if let url = URL(string: viewModel.url) {
+            
+            if let fileURL = URL(string: url.absoluteString) {
+                let fileAccessURL = fileURL.deletingLastPathComponent()
+                if let pdfData = try? Data(contentsOf: url) {
+                    webview.load(
+                        pdfData,
+                        mimeType: "application/pdf",
```
[the remark this answers is not in the document]
Comment (IvanStepanok) [reply Core/Core/View/Base/FileWebView.swift:36]: I added this part myself, it goes beyond the requirements and out of scope.❤️

File: Course/Course/Presentation/Container/CourseContainerViewModel.swift [hunk 14]
```diff
@@ -467,33 +450,222 @@ public class CourseContainerViewModel: BaseCourseViewModel {
...
             }
         } catch let error {
             if error is NoWiFiError {
                 errorMessage = CoreLocalization.Error.wifi
             }
         }
     }
-
+    
+    private func presentNoInternetAlert(sequentials: [CourseSequential]) {
```
[the remark this answers is not in the document]
Comment (IvanStepanok) [reply Course/Course/Presentation/Container/CourseContainerViewModel.swift:500]: You're right, it's a good observation! However, this alert could still be useful if the internet connection drops during the download process. The user might start downloading with a good connection, but if it's lost midway, this alert would inform them why the download stopped. It's a nice failsafe for unexpected network changes during long operations.

File: Course/Course/Presentation/Offline/OfflineView.swift [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+                Theme.Colors.background
+                    .ignoresSafeArea()
+            )
+        }
+    }
+    
+    @ViewBuilder
+    private var downloadAll: some View {
+        if viewModel.connectivity.isInternetAvaliable
+            && (viewModel.totalFilesSize - viewModel.downloadedFilesSize != 0)
```
[the remark this answers is not in the document]
Comment (IvanStepanok) [reply Course/Course/Presentation/Offline/OfflineView.swift:163]: Thank you for pointing that out; you're absolutely right. I'll make the necessary changes.

File: Course/Course/Presentation/Offline/OfflineView.swift (hunk 5 above)
Comment (IvanStepanok) [reply Course/Course/Presentation/Offline/OfflineView.swift:235]: Sure

File: Course/Course/Presentation/Unit/Subviews/OfflineContentView.swift (hunk 11 above)
Comment (IvanStepanok) [reply Course/Course/Presentation/Unit/Subviews/OfflineContentView.swift:61]: Sure, why not🫶

File: OpenEdX/AppDelegate.swift (hunk 12 above)
Comment (IvanStepanok) [reply OpenEdX/AppDelegate.swift:206]: 🥷

**Comment by IvanStepanok** [comment]
@rnr Thanks for review! All feedback addressed.

**Review by rnr — APPROVED** [review]
LGTM! Sorry for delay

**Review by volodymyr-chekyrta — APPROVED** [review]

**Pull request merged**